File: src/test/java/com/checkinx/demo2/dao/PetsRepositoryIT.java
package com.checkinx.demo2.dao;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.sql.SQLException;
import java.util.List;
import java.util.Map;

import javax.sql.DataSource;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.postgresql.jdbc.PgResultSet;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.jdbc.Sql;

import com.checkinx.AbstractIntegrationTest;
import com.checkinx.demo2.models.Pet;
import com.checkinx.utils.CheckInxAssert;
import com.checkinx.utils.CoverLevel;
import com.checkinx.utils.sql.interceptors.SqlInterceptor;
import com.checkinx.utils.sql.interceptors.postgres.PostgresInterceptor;
import com.checkinx.utils.sql.plan.parse.ExecutionPlanParser;
import com.checkinx.utils.sql.plan.parse.models.ExecutionPlan;
import com.checkinx.utils.sql.plan.parse.models.PlanNode;
import com.checkinx.utils.sql.plan.query.ExecutionPlanQuery;
import com.zaxxer.hikari.pool.HikariProxyResultSet;
import net.ttddyy.dsproxy.ExecutionInfo;
import net.ttddyy.dsproxy.QueryInfo;
import net.ttddyy.dsproxy.asserts.ProxyTestDataSource;
import net.ttddyy.dsproxy.asserts.QueryExecution;
import net.ttddyy.dsproxy.listener.QueryExecutionListener;
import net.ttddyy.dsproxy.support.ProxyDataSource;

public class PetsRepositoryIT extends AbstractIntegrationTest {

    @Autowired
    private PetsRepository repository;

    @Autowired
    private DataSource dataSource;

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private ExecutionPlanQuery executionPlanQuery;

    private SqlInterceptor sqlInterceptor;

    @Autowired
    private ExecutionPlanParser executionPlanParser;

    @Before
    public void setUp() {
        sqlInterceptor = new PostgresInterceptor((ProxyDataSource) dataSource);
        sqlInterceptor.startInterception();
    }

    @Sql("pets.sql")
    @Test
    public void testFindByNameWithProxyTestDataSourceWithJpa() {
        // ARRANGE
        final ProxyTestDataSource ds = new ProxyTestDataSource(dataSource);
        ((ProxyDataSource)dataSource).addListener(ds.getQueryExecutionFactoryListener());

        // ACT
        final List<Pet> pets = repository.findByName("Jack");

        // ASSERT
        assertEquals(1, pets.size());
        final List<QueryExecution> queryExecutions = ds.getQueryExecutions();
        assertNotNull(queryExecutions);
    }

    @Ignore
    @Sql("pets.sql")
    @Test
    public void testFindByNameWhenProxyTestDataSourceWithJdbc() {
        // ARRANGE
        final ProxyTestDataSource ds = new ProxyTestDataSource(dataSource);

        // ACT
        final List<Map<String, Object>> list = jdbcTemplate.queryForList("select * from pets");

        // ASSERT
        final String query = ds.getFirstPrepared().getQuery();
    }

    @Ignore
    @Sql("pets.sql")
    @Test
    public void testFindByNameWhenProxyDataSourceHandles() {
        // ARRANGE
        ((ProxyDataSource)dataSource).addListener(new QueryExecutionListener() {
            @Override
            public void beforeQuery(ExecutionInfo execInfo, List<QueryInfo> queryInfoList) {
                // nothing
            }

            @Override
            public void afterQuery(ExecutionInfo execInfo, List<QueryInfo> queryInfoList) {
                try {
//                    System.out.println(((HikariProxyResultSet)execInfo.getResult()).getStatement());

                    final String sql = ((HikariProxyResultSet) execInfo.getResult()).unwrap(PgResultSet.class).getStatement().toString();
//                    final List<Map<String, Object>> result = jdbcTemplate.queryForList("explain " + sql);
                    System.out.println(sql);
                }
                catch (SQLException e) {
                    e.printStackTrace();
                }
            }
        });

        // ACT
        final List<Pet> pets = repository.findByName("Jack");

        // ASSERT
        assertEquals(1, pets.size());
    }

    @Sql("pets.sql")
    @Test
    public void testFindByNameWhenNoIndex() {
        // ARRANGE

        // ACT
        final List<Pet> pets = repository.findByName("Jack");

        // ASSERT
        sqlInterceptor.stopInterception();
        assertEquals(1, sqlInterceptor.getStatements().size());

        final List<String> executionPlan = executionPlanQuery.execute(sqlInterceptor.getStatements().get(0));
        assertTrue(executionPlan.size() > 0);

        final ExecutionPlan plan = executionPlanParser.parse(executionPlan);
        assertEquals("pets pet0_ ", plan.getTable());
        assertEquals("Seq Scan", plan.getRootPlanNode().getCoverage());

//        CheckInxAssert.assertIndex(CoverLevel.ZERO, plan);

//        final List<Map<String, Object>> query = jdbcTemplate.queryForList("explain " + sqlInterceptor.getStatements().get(0));
//        assertNotNull(query);
    }

    @Sql("pets.sql")
    @Test
    public void testFindByNameWhenTextIndexedField() {
        // ARRANGE

        // ACT
        final List<Pet> pets = repository.findByLocation("Moscow");

        // ASSERT
        sqlInterceptor.stopInterception();
        assertEquals(1, sqlInterceptor.getStatements().size());

        final List<String> executionPlan = executionPlanQuery.execute(sqlInterceptor.getStatements().get(0));
        assertTrue(executionPlan.size() > 0);

        final ExecutionPlan plan = executionPlanParser.parse(executionPlan);
        assertNotNull(plan);

        final PlanNode childNode = plan.getRootPlanNode().getChildren().get(0);
        assertEquals("ix_pets_location", childNode.getTarget());
        assertEquals("Bitmap Index Scan", childNode.getCoverage());

        CheckInxAssert.assertIndex(CoverLevel.HALF, "index name", plan);
//        CheckInxAssert.assertIndex(CoverLevel.NOT_INDEX, plan);

//        final List<Map<String, Object>> query = jdbcTemplate.queryForList("explain " + sqlInterceptor.getStatements().get(0));
//        assertNotNull(query);
    }

    @Sql("pets.sql")
    @Test
    public void testFindByNameWithCheckInxAssertWhenIntIndexedField() {
        // ARRANGE

        // ACT
        final List<Pet> pets = repository.findByAge(1);

        // ASSERT
        sqlInterceptor.stopInterception();
        assertEquals(1, sqlInterceptor.getStatements().size());

        final List<String> executionPlan = executionPlanQuery.execute(sqlInterceptor.getStatements().get(0));
        assertTrue(executionPlan.size() > 0);

//        final ExecutionPlan plan = executionPlanParser.parse(executionPlan);

//        CheckInxAssert.assertIndex(CoverLevel.FULL, "index name", plan);

//        CheckInxAssert.assertIndex(CoverLevel.HALF, "index name", plan);
//        CheckInxAssert.assertIndex(CoverLevel.NOT_INDEX, plan);

//        final List<Map<String, Object>> query = jdbcTemplate.queryForList("explain " + sqlInterceptor.getStatements().get(0));
//        assertNotNull(query);
    }
}

File: src/test/java/com/checkinx/demo2/dao/PetsRepositoryIT.kt
package com.checkinx.demo2.dao

import com.checkinx.AbstractIntegrationTest
import com.checkinx.demo2.models.Pet
import com.checkinx.utils.asserts.CheckInxAssertService
import com.checkinx.utils.asserts.CoverageLevel
import com.checkinx.utils.sql.interceptors.SqlInterceptor
import com.checkinx.utils.sql.plan.parse.ExecutionPlanParser
import com.checkinx.utils.sql.plan.query.ExecutionPlanQuery
import net.ttddyy.dsproxy.asserts.ProxyTestDataSource
import net.ttddyy.dsproxy.support.ProxyDataSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.JdbcTemplate
import org.testng.AssertJUnit.*
import org.testng.annotations.BeforeClass
import org.testng.annotations.Ignore
import org.testng.annotations.Test
import java.util.*
import javax.sql.DataSource

class PetsRepositoryIT : AbstractIntegrationTest() {

    @Autowired
    private lateinit var repository: PetsRepository
    @Autowired
    private lateinit var dataSource: DataSource
    @Autowired
    private lateinit var jdbcTemplate: JdbcTemplate
    @Autowired
    private lateinit var executionPlanQuery: ExecutionPlanQuery
    @Autowired
    private lateinit var executionPlanParser: ExecutionPlanParser
    @Autowired
    private lateinit var checkInxAssertService: CheckInxAssertService
    @Autowired
    private lateinit var sqlInterceptor: SqlInterceptor

    @BeforeClass
    fun setUp() {
        // generate test data by code
        IntRange(1, 1000).forEach {
            val pet = Pet()
            pet.id = UUID.randomUUID()
            pet.age = it
            pet.location = "Saint Petersburg"
            pet.name = "Jack-$it"

            repository.save(pet)
        }
    }

    @Test
    fun testFindByNameWithProxyTestDataSourceWithJpa() {
        // ARRANGE
        val ds = ProxyTestDataSource(dataSource)
        (dataSource as ProxyDataSource).addListener(ds.queryExecutionFactoryListener)

        // ACT
        val pets = repository.findByName("Jack-100")

        // ASSERT
        assertEquals(1, pets.size.toLong())
        val queryExecutions = ds.queryExecutions
        assertNotNull(queryExecutions)
    }

    @Ignore
    @Test
    fun testFindByNameWhenProxyTestDataSourceWithJdbc() {
        // ARRANGE
        val ds = ProxyTestDataSource(dataSource)

        // ACT
        val list = jdbcTemplate.queryForList("select * from pets")

        // ASSERT
        val query = ds.firstPrepared.query
    }

    @Test
    fun testFindByNameWhenNoIndex() {
        // ARRANGE
        val name = "Jack-200"

        // ACT
        sqlInterceptor.startInterception()

        val pets = repository.findByName(name)

        sqlInterceptor.stopInterception()

        // ASSERT
        assertEquals(1, sqlInterceptor.statements.size.toLong())

        val executionPlan = executionPlanQuery.execute(sqlInterceptor.statements[0])
        assertTrue(executionPlan.isNotEmpty())

        val plan = executionPlanParser.parse(executionPlan)
        assertEquals("pets pet0_", plan.rootPlanNode.target)
        assertEquals("Seq Scan", plan.rootPlanNode.coverage)

        checkInxAssertService.assertCoverage(CoverageLevel.ZERO, "pets pet0_", plan)
    }

    // If you want to get truthful execution plan, generate enough test data
    @Test
    fun findByLocation() {
        // ARRANGE
        val location = "Moscow"

        // ACT

        // After all arrangements start interception of sql statements
        sqlInterceptor.startInterception()

        // Your investigation might be here
        val pets = repository.findByLocation(location)

        // After all investigating queries finished stop interception
        sqlInterceptor.stopInterception()

        // ASSERT

        // Here you can check how many queries were executed
        assertEquals(1, sqlInterceptor.statements.size.toLong())

        // If you want something spicy, you can parse raw plan on your own ...
        val executionPlan = executionPlanQuery.execute(sqlInterceptor.statements[0])
        assertTrue(executionPlan.isNotEmpty())

        // ... or travers plan tree ...
        val plan = executionPlanParser.parse(executionPlan)
        assertNotNull(plan)

        val rootNode = plan.rootPlanNode
        assertEquals("Bitmap Heap Scan", rootNode.coverage)
        assertEquals("pets pet0_", rootNode.target)

        // Now assert coverage is simple like never before ...
        checkInxAssertService.assertCoverage(CoverageLevel.HALF, "ix_pets_location", plan)

        // One more thing, it even could be more simple ...
        checkInxAssertService.assertCoverage(CoverageLevel.HALF, "ix_pets_location", sqlInterceptor.statements[0])

        // or if you just want to prevent "seq scan" for example, without searching concrete index
        checkInxAssertService.assertCoverage(CoverageLevel.HALF, sqlInterceptor.statements[0])
    }

    @Test
    fun testFindByNameWithCheckInxAssertWhenIntIndexedField() {
        // ARRANGE
        val age = 1

        // ACT
        sqlInterceptor.startInterception()

        val pets = repository.findByAge(age)

        sqlInterceptor.stopInterception()

        // ASSERT
        assertEquals(1, sqlInterceptor.statements.size.toLong())

        val executionPlan = executionPlanQuery.execute(sqlInterceptor.statements[0])
        assertTrue(executionPlan.isNotEmpty())

        val plan = executionPlanParser.parse(executionPlan)
        checkInxAssertService.assertCoverage(CoverageLevel.HALF, "ix_pets_age", plan)
    }

    @Test
    fun testFindByNameGivenAgeWhenIndexedFieldThenAllCoverageHalf() {
        // ARRANGE
        val age = 1

        // ACT
        sqlInterceptor.startInterception()

        val pets = repository.findByAge(age)

        sqlInterceptor.stopInterception()

        // ASSERT
        checkInxAssertService.assertCoverage(CoverageLevel.HALF, sqlInterceptor.statements[0])
    }
}